testing/src/scenario/mocking.py

```diff
@@ -379,7 +379,9 @@ def status_set(
     def juju_log(self, level: str, message: str):
         self._context.juju_log.append(JujuLogLine(level, message))
 
-    def relation_set(self, relation_id: int, key: str, value: str, is_app: bool):
+    def relation_set(
+        self, relation_id: int, data: Mapping[str, str], is_app: bool
+    ) -> None:
```

Comment:
I wonder if our/charmers lives would be easier if we made this backwards compatible. The `relation_set` call in `model.py` already passes `data` as a keyword arg, so this could be:

```suggestion
    # TODO: Remove the backwards compatibility `key` and `value` arguments and make
    # make the signature the same as in model.py once we no longer support old ops-scenario.
    def relation_set(
        self, relation_id: int, key: str, value: str, is_app: bool, data: Mapping[str, str]=None,
    ) -> None:
        if data is None:
            data = {key: value}
```

I think this would give an easier upgrade path, with the next release of `ops-scenario` supporting both old and new `ops`, and then after that transition version we can bump the required version of `ops-scenario` for `ops[testing]` to be newer and drop this.

Maybe there's a way to put in the right versions in the `ops[testing]` and `ops-scenario` dependencies that avoids this, but I suspect that may be complicated.

Reply:
We could do this, yeah. We'd need `key` and `value` to be optional though, and I guess `is_app` as well.

Reply:
Ah indeed, sorry, was just quickly writing it out and didn't think through all the specifics. I don't really like doing this, but I feel like it's probably going to be less pain to have it for a version or two.

Reply:
Tbh I'm kind of -1 on this. Wouldn't it be better to have `ops[testing]` require the latest `ops-scenario`, and have `ops-scenario` require the latest `ops`? Whenever both are being released, this means both of them have the dependency version in the `pyproject.toml` to match the release version of the other. I guess currently charmers can sometimes get away with updating one without the other, but I don't think we should try and guarantee that will work, neither in general, nor for a defined period like 2 months. If you're bumping one, you can (and should) bump the other. But won't most charmers just bump `ops` and use the version of `ops-scenario` that comes with it?

Reply:
Well, I'm pretty sure there are still charms that use `ops-scenario` specifically for tests, rather than `ops[testing]`, and so you can easily in that situation have `ops==2.20.0` and `ops-scenario==7.2.0` and that'll break running the tests but pip won't complain.

We do want people to be using `ops[testing]` and we do want people to always be on the latest version of both `ops` and `ops-scenario`, so it's not terrible if we do this. I don't think we need to guarantee that all ops 2.x will work with ops-scenario 7.x, but I'm not super keen on them being in lockstep so that you always bump both, either.

Anyway, your call here - if you'd rather bump the versions, then that's fine (we'll need to make sure we track that somewhere so it doesn't get forgotten).

Reply:
We've switched to releasing ops and ops-scenario in lockstep with exact version requirements (thanks, Tony!), so backwards compatibility code is no longer necessary here.